### ISSUE TITLE
[BUG]: Non-library functions may be incorrectly filtered out

### ISSUE THREAD

**Issue by pmor13** [compiler-explorer/compiler-explorer#6717]
### Describe the bug

Non-library functions may be incorrectly filtered out.

Function `foo` is non-library function. However, it is incorrectly filtered out.

### Steps to reproduce

Go to link https://godbolt.org/z/e1eTn3c9E. Observe that function `foo` (non-library function) is filtered out.

### Expected behavior

Function `foo` (non-library function) is not filtered out.

### Reproduction link

https://godbolt.org/z/e1eTn3c9E

### Screenshots

![image](https://github.com/user-attachments/assets/748970d8-d489-4adb-9183-f97a54d312c5)

### Operating System

Windows 10

### Browser version

Microsoft Edge Version 126.0.2592.113 (Official build) (64-bit)

**Comment by partouf** [comment]
(there's a similar report floating around our issues list, but hard to keep track of them)

This is because everything in `foo()` is marked by the compiler to belong to the `ostream` header (everything else is optimized out)
https://godbolt.org/z/873nz7cxq

**Comment by OfekShilon** [comment]
Un-checking 'Filter / comments' displays `foo`'s code. This is probably fixable.

**Comment by partouf** [comment]
No it has nothing to do with or not keeping inline code, that only applied in the case where a function is partially library code.

It has to do with the compiler has lost the connection to the original code, it starts and ends with `ostream` code:

```
foo(int*):
.Lfunc_begin0:
        .cfi_startproc
        .file   5 "/opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/15.0.0/../../../../include/c++/15.0.0" "ostream"
        .loc    5 673 2 prologue_end
        mov     rdi, qword ptr [rip + std::cerr@GOTPCREL]
.Ltmp0:
        lea     rsi, [rip + .L.str]
.Ltmp1:
        mov     edx, 22
        jmp     std::basic_ostream<char, std::char_traits<char>>& std::__ostream_insert<char, std::char_traits<char>>(std::basic_ostream<char, std::char_traits<char>>&, char const*, long)@PLT
.Ltmp3:
.Lfunc_end0:
        .size   foo(int*), .Lfunc_end0-foo(int*)
        .cfi_endproc
```

**Comment by OfekShilon** [comment]
(1) I stepped in a debugger to confirm that the `mov     rdi, qword...` line is filtered out by this exact condition -
```
 if (!keepInlineCode) { 
     continue;     <----
 }
```
(2) If this is intended behavior for such output, why does unchecking 'Filter\comments' restore the line?

**Comment by partouf** [comment]
It's probably a bug that it appears at all with the comments filter.
The behavior of the Library filter is what counts here, and it works 100% correct as intended, if you switch it off, `foo()` appears.

**Comment by OfekShilon** [comment]
@pmor13 I looked into this again: this code is undefined behavior, as the branch `*data = 7` is called only when `data` is null. 
clang reasons that the only branch with well-defined behavior is the one with ` std::cerr << ...` and so that is all the function does, and it is in fact correct to say that the only code in the function is library code.
I do see problems in our filtering code - but ultimately it does the right thing for this input. Closing for now.
